cc-proxy/src/mergeConversations.js

```diff
@@ -0,0 +1,143 @@
+ * @param {*} rsvps
+ * @returns {integer}
+ */
+function countAttendees(rsvps) {
+	const attendeeTypes = ['host', 'co-host', 'guest'];
+	const allAttendees = rsvps
+		.filter((r) => attendeeTypes.includes(r.type))
+		.map((r) => r.userUuid);
+	const uniqueAttendees = _.uniq(allAttendees);
+	return uniqueAttendees.length;
+}
+
+// compares both conversations
+// returns Uuid with more RSVPs
+async function mostRSVPs(conversationUuid1, conversationUuid2) {
+	let conversation1RSVP = await raisely(
+		{
+			method: 'GET',
+			path: `/events/${conversationUuid1}/rsvps?private=1`,
+			escalate: true,
+		},
+		req
+	);
+	let conversation2RSVP = await raisely(
+		{
+			method: 'GET',
+			path: `/events/${conversationUuid2}/rsvps?private=1`,
+			escalate: true,
+		},
+		req
+	);
+
+	let conversation1AttendeesCount = countAttendees(conversation1RSVP.data);
+	let conversation2AttendeesCount = countAttendees(conversation2RSVP.data);
+
+	return conversation1AttendeesCount >= conversation2AttendeesCount
+		? conversationUuid1
+		: conversationUuid2;
+}
+
+async function mergeConversations(req) {
+	let { conversationUuid1, conversationUuid2 } = req.body.data;
+
+	let conversation1Data = await raisely(
+		{
+			method: 'GET',
+			path: `/events/${conversationUuid1}?private=1`,
+			escalate: true,
+		},
+		req
+	);
+
+	let conversation2Data = await raisely(
+		{
+			method: 'GET',
+			path: `/events/${conversationUuid2}?private=1`,
+			escalate: true,
+		},
+		req
+	);
+
+	// Sanity check: should reject merge if conversation1Data.name !== conversation2Data.name
+	if (conversation1Data.data.name !== conversation2Data.data.name) {
+		console.error('Both conversations have the same name!');
+	}
+
+	// select which uuid to keep. Keep the uuid with the most RSVPs
+	const uuidToKeep = await mostRSVPs(conversationUuid1, conversationUuid2);
+
+	let conversationToKeep =
+		conversation1Data.data.uuid == uuidToKeep
+			? conversation1Data
+			: conversation2Data;
+	let conversationToDelete =
+		conversation1Data.data.uuid !== uuidToKeep
+			? conversation2Data
+			: conversation1Data;
+
+	// go through all the keys in the conversation to keep the
+	// and merge them into the selected conversation
+	// note that some fields only appear after they are completed (eg cash donations)
+	// (at key, if it has values, keep the field.
+	// if not, check the difference, which field to keep
+
+	for (let [key, value] of Object.entries(conversationToKeep)) {
+		if (value) {
+			return;
+		} else {
+			// not too sure about this step
+			conversationToKeep.value = conversationToDelete.value;
+		}
+	}
+
+	// point interactions to the right uuid
+	// here returns a 500 error
+	// let interactions = await raisely({
+	// 	method: 'GET',
+	// 	path: `/interactions`,
+	// 	query: {
+	// 		private: 1,
+	// 		user: conversation1Data.userUuid,
+	// 		['detail.private.conversationUuid']: conversationUuid1,
+	// 	},
+	// });
+
+	// update the conversation to keep
+	// await raisely(
```

Comment:
We'll probably want to build in some redundancy to these requests.
I can imagine in some cases we might fire off a dozen PATCH/POST/DELETE requests concurrently when we're doing this, and if one of them fails it could leave the merge in a bad state.

I'd suggest adding a `retries: 3` option on the raisely function and updating that function to repeat the request up to that many times if it gets a 500 response or a network error (but not if it gets a 400 which indicates we've sent bad data and it will never work)

Reply:
add retries option in raiselyRequest.js
based on the number of retries, perform API call in a do/while loop
check const result = request(requestOptions); (line 83)